kernel/timeout.c

```diff
@@ -253,12 +242,20 @@ void sys_clock_announce(int32_t ticks)
 	z_time_slice(ticks);
 #endif
 
-#ifdef CONFIG_SMP
-	k_spinlock_key_t ann_key = k_spin_lock(&ann_lock);
-#endif
-
 	k_spinlock_key_t key = k_spin_lock(&timeout_lock);
 
+	/* We release the lock around the callbacks below, so on SMP
+	 * systems someone might be already running the loop.  Don't
+	 * race (which will cause paralllel execution of "sequential"
+	 * timeouts and confuse apps), just increment the tick count
+	 * and return.
+	 */
+	if (IS_ENABLED(CONFIG_SMP) && announce_remaining != 0) {
+		announce_remaining += ticks;
```

Comment:
Wonder if we should use `atomic_add()` here.

Reply:
We're inside a critical section, so no need.  The spinlock is already using the atomic API, which uses a SEQ_CST builtin that will emit needed barrier instructions on architectures that require it.  There's a PR out there (maybe already merged?) that clarifies that Zephyr atomics are guaranteed to provide memory ordering.

Reply:
I considered the same thing myself so perhaps a comment here is worth adding noting atomics aren't needed, that this is already done in a critical section with what amounts to a memory barrier.